ipykernel/kernelbase.py

```diff
@@ -356,28 +361,91 @@ async def advance_eventloop():
     def _message_counter_default(self):
         return itertools.count()
 
-    async def shell_main(self):
+    async def shell_channel_thread_main(self):
+        """Main loop for shell channel thread.
+
+        Listen for incoming messages on kernel shell_socket.  For each message
+        received, extract the subshell_id from the message header and forward the
```

Comment:
Also vague question on subshell_id, is it arbitrary string and should we limit in lenght and to ascci, or something similar.

Reply:
The [JEP](https://github.com/jupyter/enhancement-proposals/blob/master/kernel-subshells/kernel-subshells.md) says that a `subshell_id` is a string, is unique within a kernel, and it is generated for you. I implemented this as `uuid.uuid4()`:
https://github.com/ianthomas23/ipykernel/blob/a708e0ee8c31a5e91175b484e76f8af48be93235/ipykernel/subshell_manager.py#L175
As it cannot be an arbitrary string I haven't put in any explicit validation as it is either recognised (corresponding to an actual subshell) or it isn't.

Reply:
Ok, I think that's fine.